packages/teams-js/src/private/externalAppAuthentication.ts

```diff
@@ -0,0 +1,329 @@
+import { authentication, FrameContexts } from '../public';
+import { errorNotSupportedOnPlatform } from '../public/constants';
+import { runtime } from '../public/runtime';
+
+/**
+ * @hidden
+ * @internal
+ * Limited to Microsoft-internal use
+ */
+export namespace externalAppAuthentication {
+  /*********** BEGIN REQUEST TYPE ************/
+  /**
+   * @hidden
+   *
+   * @internal
+   * Limited to Microsoft-internal use
+   */
+  export type OriginalRequestInfo = IQueryMessageExtensionRequest | IActionExecuteInvokeRequest;
+
+  /**
+   * @hidden
+   *
+   * @internal
+   * Limited to Microsoft-internal use
+   */
+  export interface IQueryMessageExtensionRequest {
+    requestType: OriginalRequestType.QueryMessageExtensionRequest;
+    commandId: string;
+    parameters?: {
+      name?: string;
+      value?: string;
+    }[];
+    queryOptions?: {
+      count: number;
+      skip: number;
+    };
+  }
+
+  /**
+   * @hidden
+   *
+   * @internal
+   * Limited to Microsoft-internal use
+   */
+  interface IActionExecuteInvokeRequest {
+    requestType: OriginalRequestType.ActionExecuteInvokeRequest;
+    type: string; // "invoke"
+    id: string; // "action id"
+    verb: string; // "action"
+    data: Record<string, unknown>; //object; // {}
+  }
+
+  /**
+   * @hidden
+   *
+   * @internal
+   * Limited to Microsoft-internal use
+   */
+  export enum OriginalRequestType {
+    ActionExecuteInvokeRequest = 'ActionExecuteInvokeRequest',
+    QueryMessageExtensionRequest = 'QueryMessageExtensionRequest',
+  }
+  /*********** END REQUEST TYPE ************/
+
+  /*********** BEGIN RESPONSE TYPE ************/
+  /**
+   * @hidden
+   *
+   * @internal
+   * Limited to Microsoft-internal use
+   */
+  export type IInvokeResponse = IQueryMessageExtensionResponse | IActionExecuteResponse;
+
+  /**
+   * @hidden
+   *
+   * @internal
+   * Limited to Microsoft-internal use
+   */
+  export enum InvokeResponseType {
+    ActionExecuteInvokeResponse = 'ActionExecuteInvokeResponse',
+    QueryMessageExtensionResponse = 'QueryMessageExtensionResponse',
+  }
+
+  /**
+   * @hidden
+   *
+   * @internal
+   * Limited to Microsoft-internal use
+   */
+  export interface IQueryMessageExtensionResponse {
+    responseType: InvokeResponseType.QueryMessageExtensionResponse;
+    composeExtension?: ComposeExtensionResponse;
+  }
+
+  /**
+   * @hidden
+   *
+   * @internal
+   * Limited to Microsoft-internal use
+   */
+  export interface IActionExecuteResponse {
+    responseType: InvokeResponseType.ActionExecuteInvokeResponse;
+    value: Record<string, unknown>;
+    signature?: string;
+    statusCode: number;
+    type: string;
+  }
+
+  /**
+   * @hidden
+   *
+   * @internal
+   * Limited to Microsoft-internal use
+   */
+  export type ComposeExtensionResponse = {
+    attachmentLayout: AttachmentLayout;
+    type: ComposeResultTypes;
+    attachments: QueryMessageExtensionAttachment[];
+    suggestedActions?: QueryMessageExtensionSuggestedActions;
```

Comment:
>suggestedActions [](http://example.com/codeflow?start=4&length=16)

This property has the same challenge I mentioned above for `parameters`. It seems to allow all these and I'm not sure what the semantic distinction is between these different states: undefined `suggestedActions`, defined `suggestedActions` with an undefined `actions` property, defined `suggestedActions` with a defined `actions` property that is an empty array

Reply:
@ashishsingh2105 and @susanshen96 -- can you please help clarify what valid (and invalid) input looks like here? 

Due to timelines, we can add input validation when we get more information.

Reply:
`suggestedActions` can be `null`, `undefined`, or defined `suggestedActions` with defined `actions` array.

Reply:
So to confirm: defined `suggestedActions` with undefined `actions` array is not a valid input?

Reply:
It shouldn't be.